ruby/red-arrow/benchmark/raw_records/boolean.yml

```
@@ -0,0 +1,52 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+prelude: |-
+  require 'arrow'
+  require 'faker'
+
+  state = ENV.fetch('FAKER_RANDOM_SEED', 17).to_i
+  Faker::Config.random = Random.new(state)
+
+  nrows = 1000
+  ncols = 10
+  type = :boolean
+  fields = ncols.times.map {|i| ["col_#{i}".to_sym, type] }.to_h
```

We can pass raw `Array` instead of `Arrow::BooleanArray` to `Arrow::RecordBatch`.

```suggestion
  fields = {}
  records = {}
  n_columns.times do |i|
    column_name = "column_#{i}"
    fields[column_name] = type
    records[column_name] = n_rows.times.map {Faker::Boolean.boolean}
  end
  record_batch = Arrow::RecordBatch.new(fields, records)
```

---

I replaced the record batch generation code as you showed except for DictionaryArray.
As I tried to replace the generation code for DictionaryArray, I got the following error:

```
Traceback (most recent call last):
        7: from /var/folders/84/pwfzmsk15zggy68x73f_6t7r0000gn/T/benchmark_driver-20190220-28766-12yepw8.rb:23:in `<main>'
        6: from /Users/mrkn/src/github.com/apache/arrow/ruby/red-arrow/lib/arrow/record-batch.rb:31:in `new'
        5: from /Users/mrkn/src/github.com/apache/arrow/ruby/red-arrow/lib/arrow/record-batch-builder.rb:23:in `build'
        4: from /Users/mrkn/src/github.com/apache/arrow/ruby/red-arrow/lib/arrow/record-batch-builder.rb:23:in `new'
        3: from /Users/mrkn/src/github.com/apache/arrow/ruby/red-arrow/lib/arrow/record-batch-builder.rb:35:in `initialize'
        2: from /Users/mrkn/.rbenv/versions/2.6.0/lib/ruby/gems/2.6.0/gems/gobject-introspection-3.3.1/lib/gobject-introspection/loader.rb:328:in `block in load_constructor_infos'
        1: from /Users/mrkn/.rbenv/versions/2.6.0/lib/ruby/gems/2.6.0/gems/gobject-introspection-3.3.1/lib/gobject-introspection/loader.rb:317:in `block (2 levels) in load_constructor_infos'
/Users/mrkn/.rbenv/versions/2.6.0/lib/ruby/gems/2.6.0/gems/gobject-introspection-3.3.1/lib/gobject-introspection/loader.rb:317:in `invoke': [record-batch-builder][new]: NotImplemented: MakeBuilder: cannot construct builder for type dictionary<values=string, indices=int8, ordered=1> (Arrow::Error::NotImplemented)
```

The modification I tried is given below:

```diff
diff --git a/ruby/red-arrow/benchmark/raw_records/dict.yml b/ruby/red-arrow/benchmark/raw_records/dict.yml
index 47844ab0..562f6e2f 100644
--- a/ruby/red-arrow/benchmark/raw_records/dict.yml
+++ b/ruby/red-arrow/benchmark/raw_records/dict.yml
@@ -29,19 +29,16 @@ prelude: |-
   )
   type = Arrow::DictionaryDataType.new(:int8, dict_vocab, true)

-  fields = n_columns.times.map {|i| ["column_#{i}".to_sym, type] }.to_h
-  schema = Arrow::Schema.new(**fields)
-  arrays = n_columns.times.map do
-    Arrow::DictionaryArray.new(
-      type,
-      Arrow::Int8Array.new(
-        n_rows.times.map {
-          Faker::Number.within(0 ... dict_vocab.length)
-        }
-      )
-    )
+  fields = {}
+  arrays = {}
+  n_columns.times do |i|
+    column_name = "column_#{i}"
+    fields[column_name] = type
+    arrays[column_name] = n_rows.times.map {
+      Faker::Number.within(0 ... dict_vocab.length)
+    }
   end
-  record_batch = Arrow::RecordBatch.new(schema, n_rows, arrays)
+  record_batch = Arrow::RecordBatch.new(fields, arrays)

   def pure_ruby_raw_records(record_batch)
     n_rows = record_batch.n_rows
```

---

It seems that we need to add `case Ttype::DICTIONARY` clause that uses `DictionaryBuilder` in `MakeBuilder()`:

  * https://github.com/apache/arrow/blob/master/cpp/src/arrow/builder.cc#L95
  * https://github.com/apache/arrow/blob/master/cpp/src/arrow/array/builder_dict.h

It should be done in a separated issue. Could you open a JIRA issue?

---

I see. I'll do that later.

---

I made the issue for MakeBuilder.
I rewrote the benchmark for DictionaryArray after resolving that issue.

---

@kou I understand that it is difficult to let MakeBuilder support to create DictionaryBuilder.
The reason is MakeBuilder requires a value type of the resultant Array, while DictionaryBuildr also builds the value type of the resultant DictionaryArray.

---

OK. I also look into this.